### PR TITLE
Implement complex norm in optimizers

### PR DESCRIPTION
**Pull request by wdphy16**
As proposed in #196, I reviewed all transformations in `transform.py` and optimizers in `alias.py`, and replaced `x**2` with `(x.conj() * x).real` to support complex numbers.

I think we need to double check the correctness of all changes. What tests do you think we should add?

I'm not sure how to generalize the transform `sm3` to complex numbers. The optimizers `fromage` and `noisy_sgd` should be working, but in the tests I can't find a learning rate to make them work well.

### CODE REVIEW

**Comment by wdphy16** [comment]
I rewrote `AliasTest.test_optimization` in `alias_test.py` to test for complex parameters on two target functions. By the way, I added the missing tests for all optimizers in `alias.py`.

**Comment by PhilipVinc** [comment]
this should also use the complex norm in gradient clipping. See #287

**Comment by wdphy16** [comment]
Now I've also implemented complex gradient clipping in this PR (and closed #161), and I hope it doesn't make this PR look too large.

**Comment by PhilipVinc** [comment]
I'd suggest renaming `abs_sqr` to `abs2`, which is the name used in [Scipy](https://github.com/scipy/scipy/pull/15390), [C++ Eigen](https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html), [Julia](https://docs.julialang.org/en/v1/base/math/#Base.abs2) ...

**Comment by PhilipVinc** [comment]
@mkunesch If you could spare the time, could we get a review on this PR?

**Comment by wdphy16** [comment]
> I'd suggest renaming `abs_sqr` to `abs2`, which is the name used in [Scipy](https://github.com/scipy/scipy/pull/15390), [C++ Eigen](https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html), [Julia](https://docs.julialang.org/en/v1/base/math/#Base.abs2) ...

Yes I agree

File: optax/_src/transform.py [hunk 1]
```diff
@@ -76,17 +76,39 @@ def update_fn(updates, state, params=None):
 
 
 def _update_moment(updates, moments, decay, order):
-  """Compute the exponential moving average of the `order-th` moment."""
+  """Compute the exponential moving average of the `order`-th moment."""
   return jax.tree_multimap(
       lambda g, t: (1 - decay) * (g ** order) + decay * t, updates, moments)
 
 
+def _update_moment_norm(updates, moments, decay, order):
```
Comment (hbq1) [inline optax/_src/transform.py:84]: I find it slightly confusing because "norm" can be interpreted as a vector/matrix norm. Perhaps `_update_moment_per_elem_norm` would fit better?

File: optax/_src/transform.py [hunk 2]
```diff
@@ -483,9 +505,10 @@ def init_fn(params):
   def update_fn(updates, state, params=None):
     del params
     mu = _update_moment(updates, state.mu, b1, 1)
-    signed_sq = jax.tree_multimap(
-        lambda g, v: jnp.sign(v - g**2)*g**2, updates, state.nu)
-    nu = _update_moment(signed_sq, state.nu, b2, 2)
+    nu = jax.tree_multimap(
+        lambda g, v: v - (1 - b2) * jnp.sign(v - numerics.abs2(g)) *
+            numerics.abs2(g),
+        updates, state.nu)
```
Comment (hbq1) [inline optax/_src/transform.py:511]: Nice catch! Can you send a separate PR with this fix, please?

File: optax/_src/numerics.py [hunk 3]
```diff
@@ -23,6 +23,13 @@
 # TODO(jscholz) Promote these functions to jax core lib?
 
 
+def abs2(x: chex.Array) -> chex.Array:
```
Comment (hbq1) [inline optax/_src/numerics.py:26]: Consider making it a part of public API (i.e. adding it to __init__.py)

File: optax/_src/alias_test.py [hunk 4]
```diff
@@ -22,80 +22,85 @@
 import jax.numpy as jnp
 
 from optax._src import alias
+from optax._src import numerics
 from optax._src import update
 
 
-class AliasTest(chex.TestCase):
+def _setup_parabel(dtype):
```
Comment (hbq1) [inline optax/_src/alias_test.py:29]: Parabole/parabola?

File: optax/_src/alias_test.py [hunk 5]
```diff
@@ -22,80 +22,85 @@
...
+          dict(opt_name='sm3', opt=lambda: alias.sm3(1.0)),
+          dict(opt_name='yogi', opt=lambda: alias.yogi(1e-1)),
+          dict(opt_name='dpsgd',
+              opt=lambda: alias.dpsgd(1e-3, 10.0, 0.001, 0, 0.2)),
+      ),
+      target=(_setup_parabel, _setup_rosenbrock),
+      dtype=(jnp.float32, jnp.complex64),
+  )
+  def test_optimization(self, opt_name, opt, target, dtype):
+    if (opt_name in ['fromage', 'noisy_sgd', 'sm3']
```
Comment (hbq1) [inline optax/_src/alias_test.py:90]: Please use a tuple here (as per google pystyle guide)

File: optax/_src/alias_test.py [hunk 6]
```diff
@@ -22,80 +22,85 @@
...
+        and jnp.iscomplexobj(dtype)):
+      raise absltest.SkipTest(
+          'This optimizer does not support complex parameters.')
```
Comment (hbq1) [inline optax/_src/alias_test.py:93]: f'{opt_name} ...' instead of 'This optimizer' ?

File: optax/_src/alias_test.py [hunk 7]
```diff
@@ -22,80 +22,85 @@
...
 
-    a = 1.0
-    b = 100.0
-    initial_params = jnp.array([0.0, 0.0])
-    final_params = jnp.array([a, a**2])
 
-    @jax.grad
-    def get_updates(params):
-      return (a - params[0])**2 + b * (params[1] - params[0]**2)**2
+class AliasTest(chex.TestCase):
```
Comment (hbq1) [inline optax/_src/alias_test.py:63]: (nit, feel free to ignore this comment) 
This test derives from chex.TestCase but doesn't use variants which can improve the coverage: https://github.com/deepmind/chex#test-variants-variantspy

**Review by hbq1 — COMMENTED** [review]
Nice work! Thanks a lot for putting this together.

**Comment by zerothi** [comment]
> I'd suggest renaming `abs_sqr` to `abs2`, which is the name used in [Scipy](https://github.com/scipy/scipy/pull/15390), [C++ Eigen](https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html), [Julia](https://docs.julialang.org/en/v1/base/math/#Base.abs2) ...

In `scipy` it has been decided to name it `abs_sq`, not `abs2` FYI.

**Comment by hbq1** [comment]
> > I'd suggest renaming `abs_sqr` to `abs2`, which is the name used in [Scipy](https://github.com/scipy/scipy/pull/15390), [C++ Eigen](https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html), [Julia](https://docs.julialang.org/en/v1/base/math/#Base.abs2) ...
> 
> In `scipy` it has been decided to name it `abs_sq`, not `abs2` FYI.

@wdphy16 could you update the name to `abs_sq` for (forward?) consistency with scipy, please?

**Comment by wdphy16** [comment]
> @wdphy16 could you update the name to `abs_sq` for (forward?) consistency with scipy, please?

Done that

**Comment by PhilipVinc** [comment]
@hbq1 After you merge this PR and #288 could you also tag a new release?

**Review by hbq1 — APPROVED** [review]

**Comment by hbq1** [comment]
@PhilipVinc we'll release a new version early next week if you don't mind (releasing something Friday evening might turn into a restless "weekend" ;) )

**Comment by wdphy16** [comment]
It's ok, thank you!

**Review by wdphy16 — COMMENTED** [review]
Sorry that I forgot to submit the reviews before, and now I just want to keep them for future reference
